### PR TITLE
Update LibTorch backend to new request / response flow

### PR DESCRIPTION
**Pull request by CoderHam**

### CODE REVIEW

File: src/backends/pytorch/libtorch_backend.cc [hunk 1]
```diff
@@ -357,71 +354,160 @@ LibTorchBackend::Context::ValidateOutputs(
...
+    if (!torch_dtype.first) {
+      return Status(
+          Status::Code::INTERNAL, "Failed to convert DataType '" +
+                                      DataType_Name(datatype) +
+                                      "' to Torch datatype");
+    }
+
+    // Checked at initialization time to make sure that STRING is not
+    // being used for an input, so can just assume fixed-sized here.
+    size_t total_element_cnt = 1;
```
Comment (deadeyegoodwin) [inline src/backends/pytorch/libtorch_backend.cc:421]: Use GetElementCount ?

File: src/backends/pytorch/libtorch_backend.cc [hunk 2]
```diff
@@ -357,71 +354,160 @@ LibTorchBackend::Context::ValidateOutputs(
...
+    }
+
+    const size_t total_byte_size =
+        total_element_cnt * GetDataTypeByteSize(datatype);
+
+    // The entire input tensor must be delivered as a single
+    // contiguous chunk so create a buffer large enough to hold the
+    // entire dynamic batched input.
+    input_buffers->emplace_back(new AllocatedMemory(
+        GetByteSize(datatype, batchn_shape),
```
Comment (deadeyegoodwin) [inline src/backends/pytorch/libtorch_backend.cc:433]: Didn't you just calculate total_byte_size above? Do you use total_element_cnt for something else? If not why not just use GetByteSize to set total_byte_size able?

File: src/backends/pytorch/libtorch_backend.cc [hunk 3]
```diff
@@ -357,71 +354,160 @@ LibTorchBackend::Context::ValidateOutputs(
...
+        (gpu_device_ == NO_GPU_DEVICE) ? TRITONSERVER_MEMORY_CPU_PINNED
+                                       : TRITONSERVER_MEMORY_GPU,
+        (gpu_device_ == NO_GPU_DEVICE) ? 0 : gpu_device_));
+
+    TRITONSERVER_MemoryType memory_type;
+    int64_t memory_type_id;
+    auto input_buffer =
+        input_buffers->back()->MutableBuffer(&memory_type, &memory_type_id);
```
Comment (deadeyegoodwin) [inline src/backends/pytorch/libtorch_backend.cc:441]: I don't see where you used collector to actually gather the inputs into this input_buffer. Did this pass L0_infer?

File: src/backends/pytorch/libtorch_backend.cc [hunk 4]
```diff
@@ -460,324 +546,226 @@ LibTorchBackend::Context::GetOutputTensor(
...
-  // they should be processed then to initialze
-  // input_index_map_. Since they are not we do it here for every
-  // request which is unnecessary perf overhead.
-  bool cuda_copy = false;
-  for (const auto& pr : repr_input_request->ImmutableInputs()) {
-    const InferenceRequest::Input* input = pr.second;
-    const std::string& name = input->Name();
-    int ip_index;
+  // Collect the request inputs into contiguous input tensors.
+  // FIXME override inputs from controls should be known from the model
```
Comment (deadeyegoodwin) [inline src/backends/pytorch/libtorch_backend.cc:645]: Is this comment still true?

File: src/backends/pytorch/libtorch_backend.cc [hunk 5]
```diff
@@ -460,324 +546,226 @@ LibTorchBackend::Context::GetOutputTensor(
...
   // Run...
-  RETURN_IF_ERROR(Execute(&inputs_, &outputs_));
-
-#ifdef TRTIS_ENABLE_STATS
-  for (auto& payload : *payloads) {
-    if (payload.stats_ != nullptr) {
-      payload.stats_->CaptureTimestamp(
-          ModelInferStats::TimestampKind::kComputeOutputStart);
+  Status status = Execute(&inputs_, &outputs_);
+  if (!status.IsOk()) {
```
Comment (deadeyegoodwin) [inline src/backends/pytorch/libtorch_backend.cc:682]: Use FAIL_ALL_AND_RETURN_IF_ERROR macro

File: src/backends/tensorflow/base_backend.cc [hunk 6]
```diff
@@ -459,7 +459,7 @@ SetStringInputTensor(
   // copy it there.
   auto buffer_memory_type = TRITONSERVER_MEMORY_CPU_PINNED;
   int64_t buffer_memory_type_id = 0;
-  const char* content;
+  const char* content = nullptr;
```
Comment (deadeyegoodwin) [inline src/backends/tensorflow/base_backend.cc:462]: Did you find a case where this was necessary?

File: src/backends/pytorch/libtorch_backend.cc (hunk 3 above)
Comment (CoderHam) [reply src/backends/pytorch/libtorch_backend.cc:428]: The lines were removed during rebase. Fixed now

File: src/backends/tensorflow/base_backend.cc (hunk 6 above)
Comment (CoderHam) [reply src/backends/tensorflow/base_backend.cc:462]: The compiler throws a warning regarding it remaining initialized

File: src/backends/pytorch/libtorch_backend.cc (hunk 4 above)
Comment (CoderHam) [reply src/backends/pytorch/libtorch_backend.cc]: No. Validating Control Inputs during model loading now

File: src/backends/pytorch/libtorch_backend.cc [hunk 7]
```diff
@@ -460,324 +546,215 @@ LibTorchBackend::Context::GetOutputTensor(
...
+  FAIL_ALL_AND_RETURN_IF_ERROR(
+      requests, responses, metric_reporter_.get(),
+      ReadOutputTensors(
+          base, total_batch_size, requests, &responses, &outputs_,
+          &output_index_map_),
+      "error sending LibTorch response");
 
 #ifdef TRITON_ENABLE_GPU
   if (cuda_copy) {
     cudaStreamSynchronize(stream_);
```
Comment (GuanLuo) [inline src/backends/pytorch/libtorch_backend.cc:712]: You don't need to synchronize here, already done in ReadOutputTensors

Comment (CoderHam) [reply src/backends/pytorch/libtorch_backend.cc]: Removed

File: src/backends/pytorch/libtorch_backend.cc [hunk 8]
```diff
@@ -30,6 +30,7 @@
 #include <exception>
 #include <memory>
 #include "src/core/constants.h"
+#include "src/core/cuda_utils.h"
```
Comment (GuanLuo) [inline src/backends/pytorch/libtorch_backend.cc:33]: Do you need this include?

Comment (CoderHam) [reply src/backends/pytorch/libtorch_backend.cc]: Also removed from caffe2 backend as it was no longer necessary

**Comment by CoderHam** [comment]
Yes I did

**Comment by deadeyegoodwin** [comment]
@GuanLuo since you had some comments earlier can you re-review. If you are ok with it I will merge.

**Comment by GuanLuo** [comment]
You can merge it

**Pull request merged**